### PR TITLE
HHH-17275: Fix NPE in BooleanJavaType for converter returning NULL

### PR DESCRIPTION
**Pull request by datazuul**
Can handle NULL for a boolean value (e.g. NULL relational value for java boolean "false")

https://hibernate.atlassian.net/browse/HHH-17275

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/dialect/Dialect.java [hunk 1]
```diff
@@ -807,19 +807,29 @@ public String[] getDropEnumTypeCommand(Class<? extends Enum<?>> enumType) {
...
+				nullIsValid = true;
+				continue;
+			}
 			check.append( separator ).append('\'').append( value ).append('\'');
 			separator = ",";
 		}
-		return check.append( ')' ).toString();
+		check.append( ')' );
+		if ( nullIsValid ) {
+			check.append( " OR " ).append(columnName).append( " IS NULL" );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/dialect/Dialect.java:830]:
```suggestion
			check.append( " or " ).append( columnName ).append( " is null" );
```

File: hibernate-core/src/main/java/org/hibernate/type/descriptor/java/BooleanJavaType.java [hunk 2]
```diff
@@ -192,8 +192,8 @@ public String getCheckCondition(String columnName, JdbcType jdbcType, BasicValue
 				BasicValueConverter<Boolean, ?> stringConverter =
 						(BasicValueConverter<Boolean, ?>) converter;
 				String[] values = new String[] {
-						stringConverter.toRelationalValue(false).toString(),
-						stringConverter.toRelationalValue(true).toString()
+						stringConverter.toRelationalValue(false) != null ? stringConverter.toRelationalValue(false).toString() : null,
+						stringConverter.toRelationalValue(true) != null ? stringConverter.toRelationalValue(true).toString() : null
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/type/descriptor/java/BooleanJavaType.java:196]:
```suggestion
				final Object falseValue = stringConverter.toRelationalValue( false );
				final Object trueValue = stringConverter.toRelationalValue( true );
				final String[] values = new String[] {
						falseValue != null ? falseValue.toString() : null,
						trueValue != null ? trueValue.toString() : null
```

**Review by beikov — CHANGES_REQUESTED** [review]
Thanks for your work. I just have some minor change requests :)

**Comment by datazuul** [comment]
To make the changes complete and fix the same possible NPE for Number typed Boolean handling, code and tests added for Numbers, too.

File: hibernate-core/src/main/java/org/hibernate/dialect/Dialect.java [hunk 3]
```diff
@@ -843,14 +853,39 @@ public String getCheckCondition(String columnName, long min, long max) {
 	 * @return a SQL expression that will occur in a {@code check} constraint
 	 */
 	public String getCheckCondition(String columnName, long[] values) {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/dialect/Dialect.java:855]: Please deprecate this method for removal and point to the new method.

**Comment by beikov** [comment]
Looks good. Also, please squash the commits and rebase instead of merging main.

File: hibernate-core/src/main/java/org/hibernate/dialect/Dialect.java (hunk 3 above)
Comment (datazuul) [reply hibernate-core/src/main/java/org/hibernate/dialect/Dialect.java:857]: done

**Review by beikov — APPROVED** [review]

**Pull request merged**